File: frontend/lib/data/models/pet.dart
import 'package:equatable/equatable.dart';

/// The household's shared pet (PDR-001). `hunger`/`mood` arrive already
/// decayed to "now" by the backend (economy.service.ts's computeDecay) —
/// this model just carries whatever the server sent, it never computes
/// decay itself.
class Pet extends Equatable {
  final String id;
  final String householdId;
  final String species; // cat | dog
  final String name;
  final num hunger;
  final num mood;
  final DateTime? lastFedAt;
  final DateTime? lastPlayedAt;
  final List<String> cosmetics;
  final String? activeCosmetic;

  const Pet({
    required this.id,
    required this.householdId,
    required this.species,
    required this.name,
    this.hunger = 80,
    this.mood = 80,
    this.lastFedAt,
    this.lastPlayedAt,
    this.cosmetics = const [],
    this.activeCosmetic,
  });

  factory Pet.fromJson(Map<String, dynamic> json) {
    return Pet(
      id: (json['id'] ?? json['_id'] ?? '').toString(),
      householdId: (json['householdId'] ?? '').toString(),
      species: (json['species'] ?? 'cat') as String,
      name: (json['name'] ?? '') as String,
      hunger: (json['hunger'] ?? 80) as num,
      mood: (json['mood'] ?? 80) as num,
      lastFedAt: json['lastFedAt'] != null
          ? DateTime.tryParse(json['lastFedAt'].toString())
          : null,
      lastPlayedAt: json['lastPlayedAt'] != null
          ? DateTime.tryParse(json['lastPlayedAt'].toString())
          : null,
      cosmetics: (json['cosmetics'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          const [],
      activeCosmetic: json['activeCosmetic'] as String?,
    );
  }

  Pet copyWith({
    num? hunger,
    num? mood,
    DateTime? lastFedAt,
    DateTime? lastPlayedAt,
    List<String>? cosmetics,
    String? activeCosmetic,
  }) {
    return Pet(
      id: id,
      householdId: householdId,
      species: species,
      name: name,
      hunger: hunger ?? this.hunger,
      mood: mood ?? this.mood,
      lastFedAt: lastFedAt ?? this.lastFedAt,
      lastPlayedAt: lastPlayedAt ?? this.lastPlayedAt,
      cosmetics: cosmetics ?? this.cosmetics,
      activeCosmetic: activeCosmetic ?? this.activeCosmetic,
    );
  }

  @override
  List<Object?> get props => [
        id,
        householdId,
        species,
        name,
        hunger,
        mood,
        lastFedAt,
        lastPlayedAt,
        cosmetics,
        activeCosmetic,
      ];
}

/// A household's in-flight adoption proposal (PDR-001 A2/A3): one member
/// starts it, a DIFFERENT member must confirm. Mirrors backend
/// AdoptionRequest — see GET/POST/DELETE .../pet/adopt.
class AdoptionRequest extends Equatable {
  final String id;
  final String householdId;
  final String species;
  final String name;
  final String requestedBy;

  const AdoptionRequest({
    required this.id,
    required this.householdId,
    required this.species,
    required this.name,
    required this.requestedBy,
  });

  factory AdoptionRequest.fromJson(Map<String, dynamic> json) {
    return AdoptionRequest(
      id: (json['id'] ?? json['_id'] ?? '').toString(),
      householdId: (json['householdId'] ?? '').toString(),
      species: (json['species'] ?? 'cat') as String,
      name: (json['name'] ?? '') as String,
      requestedBy: (json['requestedBy'] ?? '').toString(),
    );
  }

  @override
  List<Object?> get props => [id, householdId, species, name, requestedBy];
}

File: backend/src/models/Pet.ts
import { Schema, model, Document, Types } from 'mongoose';
import { jsonSchemaOptions } from '../utils/toJSON';

export type PetSpecies = 'cat' | 'dog';

/**
 * The household's shared pet (PDR-001 Fase A). Exactly one per household —
 * adoption is a household-level decision, not a per-user one.
 *
 * `hunger`/`mood` are the values as of the last write (feed/play/adoption);
 * they are NOT decayed in place on a timer. Reads must run them through
 * economy.service.ts's `computeDecay` to get the current, time-adjusted
 * value — see that function's doc comment for why decay is lazy.
 */
export interface IPet extends Document {
  _id: Types.ObjectId;
  householdId: Types.ObjectId;
  species: PetSpecies;
  name: string;
  adoptedAt: Date;
  adoptedBy: Types.ObjectId;
  hunger: number;
  mood: number;
  lastFedAt?: Date;
  lastPlayedAt?: Date;
  cosmetics: string[];
  activeCosmetic?: string;
  createdAt: Date;
  updatedAt: Date;
}

const petSchema = new Schema<IPet>(
  {
    householdId: {
      type: Schema.Types.ObjectId,
      ref: 'Household',
      required: true,
      unique: true,
      index: true,
    },
    species: { type: String, enum: ['cat', 'dog'], required: true },
    name: { type: String, required: true, trim: true, minlength: 1, maxlength: 50 },
    adoptedAt: { type: Date, required: true, default: Date.now },
    adoptedBy: { type: Schema.Types.ObjectId, ref: 'User', required: true },
    hunger: { type: Number, min: 0, max: 100, default: 80 },
    mood: { type: Number, min: 0, max: 100, default: 80 },
    lastFedAt: { type: Date },
    lastPlayedAt: { type: Date },
    cosmetics: { type: [String], default: [] },
    activeCosmetic: { type: String },
  },
  { timestamps: true, ...jsonSchemaOptions },
);

export const PetModel = model<IPet>('Pet', petSchema);
